RooFit::MultiProcess & TestStatistics part 3: Minuit2::NumericalDerivator

egpbos opened the pull request:
This PR introduces the NumericalDerivatorMinuit2 class, based on original code by @lmoneta (https://github.com/lmoneta/root/blob/lvmini/math/mathcore/src/NumericalDerivator.cxx). It replicates the Minuit2 numerical derivator calculations, but in a a stand alone class. This allows it to be used from RooFit to manually calculate gradients outside of Minuit2, e.g. in a parallelized way.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

egpbos commented:
Update: I have moved `NumericalDerivatorMinuit2` from roofitcore into Minuit2 and, of course, renamed it into `Minuit2::NumericalDerivator`.

Originally, @lmoneta's NumericalDerivator was in mathcore. However, I think it fits well in Minuit2 as well, since the algorithms exactly try to mimic Minuit2's. Also, I have added some depedencies to Minuit2 parameter transformations that I think make it an even more logical fit in Minuit2 than to have to make mathcore depend on Minuit2. I guess in the end it's an arbitrary choice.

lmoneta commented:
@phsft-bot build

lmoneta commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,119 @@
+                             bool always_exactly_mimic_minuit2 = true);
+   virtual ~NumericalDerivator();
+
+   void setup_differentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *cx,
```
Can you rename to `SetupDifferentiate` to be consistent with ROOT naming conventions ?

lmoneta commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,119 @@
+                 const std::vector<DerivatorElement>& previous_gradient);
+
+   DerivatorElement
+   partial_derivative(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
```
Same here and rename to "PartialDerivative" ?

lmoneta commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,119 @@
+   DerivatorElement
+   partial_derivative(const ROOT::Math::IBaseFunctionMultiDim *function, const double *x,
+                      const std::vector<ROOT::Fit::ParameterSettings> &parameters, unsigned int i_component, DerivatorElement previous);
+   DerivatorElement fast_partial_derivative(const ROOT::Math::IBaseFunctionMultiDim *function,
```
Here `FastPartiallDerivative`

lmoneta commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,119 @@
+   ROOT::Minuit2::SqrtUpParameterTransformation fUpperLimTrafo;
+   ROOT::Minuit2::SqrtLowParameterTransformation fLowerLimTrafo;
+
+private:
```
also here use ROOT convention for data members

`_always_exactly_mimic_minuit2` -> `fAlways_exactly_mimic_minuit2`

lmoneta commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,119 @@
+   double fStepTolerance = 0.5;
+   double fGradTolerance = 0.1;
+   unsigned int fNCycles = 2;
+   double Up = 1;
```
Up -> fUp

lmoneta commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,119 @@
+   double Up = 1;
+   double fVal = 0;
+
+   std::vector<double> vx, vx_external;
```
vx, vx_external -> >fVx, fVx_external

lmoneta replied:
and same for those two data members below

lmoneta reviewed (dismissed):
It looks to good to me. For clarity of the code, it would be nice to follow the ROOT coding convention for member function names and data member names, which are different than the RooFit ones. 
It would be nice also to add a direct unit test of the class, but this could be eventually a separate PR

egpbos commented:
Thanks for the suggestions @lmoneta! I have applied them all, plus one extra, namely `GetFValue()`, which I think should just be `GetValue()` (I probably just verbatim copied the `f` from the ROOT-style named `fVal`).

I also removed the `set_step_tolerance` etc functions, they were duplicates of already existing `SetStepTolerance` in the correct coding style, I must have overlooked those.

Finally, I deleted the empty virtual destructor, I don't think that had any use, so now the class is not virtual anymore, which may give some micro-performance gain.

A unit test would be good, indeed. However, #8596 does provide test coverage of the class, since RooGradMinimizerFcn uses NumericalDerivator. I hope that this is sufficient for the time being and we can indeed add unit tests in a later PR, as you suggest.

hageboeck commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,114 @@
+                           const std::vector<ROOT::Fit::ParameterSettings> &parameters,
+                           std::vector<DerivatorElement> &gradient);
+
+private:
+   double fStepTolerance = 0.5;
+   double fGradTolerance = 0.1;
+   unsigned int fNCycles = 2;
+   double fUp = 1;
+   double fVal = 0;
+
+   std::vector<double> fVx, fVxExternal;
+   double fDfmin;
+   double fVrysml;
+
+   // MODIFIED: Minuit2 determines machine precision in a slightly different way than
+   // std::numeric_limits<double>::epsilon()). We go with the Minuit2 one.
+   ROOT::Minuit2::MnMachinePrecision fPrecision;
+
+   ROOT::Minuit2::SinParameterTransformation fDoubleLimTrafo;
+   ROOT::Minuit2::SqrtUpParameterTransformation fUpperLimTrafo;
+   ROOT::Minuit2::SqrtLowParameterTransformation fLowerLimTrafo;
+
+private:
+   bool fAlwaysExactlyMimicMinuit2;
+
+public:
+   bool AlwaysExactlyMimicMinuit2() const;
+   void SetAlwaysExactlyMimicMinuit2(bool flag);
+
+private:
+   std::vector<double> fVxFValCache;
```
I grouped members according to visibility and size, so the class instances will use less space in RAM.
```suggestion
   bool AlwaysExactlyMimicMinuit2() const;
   void SetAlwaysExactlyMimicMinuit2(bool flag);
   
private:
   double fStepTolerance = 0.5;
   double fGradTolerance = 0.1;
   double fUp = 1;
   double fVal = 0;

   std::vector<double> fVx, fVxExternal;
   double fDfmin;
   double fVrysml;

   // MODIFIED: Minuit2 determines machine precision in a slightly different way than
   // std::numeric_limits<double>::epsilon()). We go with the Minuit2 one.
   ROOT::Minuit2::MnMachinePrecision fPrecision;

   ROOT::Minuit2::SinParameterTransformation fDoubleLimTrafo;
   ROOT::Minuit2::SqrtUpParameterTransformation fUpperLimTrafo;
   ROOT::Minuit2::SqrtLowParameterTransformation fLowerLimTrafo;

   unsigned int fNCycles = 2;
   bool fAlwaysExactlyMimicMinuit2;

   std::vector<double> fVxFValCache;
```

guitargeek replied:
I would also put the two vectors in separate lines to make the code more readable (it's easy to miss there are actually two `std::vector` data members when you write `std::vector<double> fVx, fVxExternal;`):
```
std::vector<double> fVx;
std::vector<double> fVxExternal;
```

hageboeck replied:
Three, actually! Suggestion updated! 👍

hageboeck commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+ * Copyright (c) 2013 , LCG ROOT MathLib Team                         *
+ *                                                                    *
+ **********************************************************************/
+/*
+ * NumericalDerivator.cxx
+ *
```
In the current version, doxygen won't generate any documentation. 🙁 
```suggestion
/**
 * \class NumericalDerivator
 *
```

hageboeck commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+{
+}
+
+void NumericalDerivator::SetStepTolerance(double value)
+{
+   fStepTolerance = value;
+}
+
+void NumericalDerivator::SetGradTolerance(double value)
+{
+   fGradTolerance = value;
+}
+
+void NumericalDerivator::SetNCycles(unsigned int value)
+{
+   fNCycles = value;
+}
+
+void NumericalDerivator::SetErrorLevel(double value)
+{
+   fUp = value;
+}
```
These might be inlined, and one-line `///` comments could be added to help users use this class.

egpbos replied:
Maybe @lmoneta or you or someone else could later on add some helpful comments. The algorithm here is a replica of the `Numerical2PGradientCalculator`. I understand more or less how it works, but not necessarily why all parameters are the way they are. Someone like @lmoneta might be better able to do this.

hageboeck commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+   fUp = value;
+}
+
+// This function sets internal state based on input parameters. This state
+// setup is used in the actual (partial) derivative calculations.
```
Doxygen only reads three `/` comments.
```suggestion
/// This function sets internal state based on input parameters. This state
/// setup is used in the actual (partial) derivative calculations.
```

hageboeck commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+}
+
+// MODIFIED:
+// This function was not implemented as in Minuit2. Now it copies the behavior
+// of InitialGradientCalculator. See https://github.com/roofit-dev/root/issues/10
```
```suggestion
/// This function was not implemented as in Minuit2. Now it copies the behavior
/// of InitialGradientCalculator. See https://github.com/roofit-dev/root/issues/10
```

hageboeck reviewed (commented):
I have a suggestion for a detail optimisation (padding bytes in the class structure), readability stuff and documentation.
In general it would be good to have *every* function documented with `///`, but I understand that some functions have self-explaining names and it's more work.

guitargeek commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+{
+}
+
+// deep copy constructor
+NumericalDerivator::NumericalDerivator(const NumericalDerivator &other)
+   : fStepTolerance(other.fStepTolerance), fGradTolerance(other.fGradTolerance), fNCycles(other.fNCycles),
+     fUp(other.fUp), fVal(other.fVal), fVx(other.fVx), fVxExternal(other.fVxExternal), fDfmin(other.fDfmin),
+     fVrysml(other.fVrysml), fPrecision(other.fPrecision), fAlwaysExactlyMimicMinuit2(other.fAlwaysExactlyMimicMinuit2),
+     fVxFValCache(other.fVxFValCache)
+{
+}
```
Is it intentional that the copy constructor doesn't copy the `fDoubleLimTrafo`, `fUpperLimTrafo`, and `fLowerLimTrafo` members? Why not copy them too, and then you can even use the `default` copy constructor? That's less error prone I think.

guitargeek replied:
Maybe if there is a reason for not copying the three members, it should be documented in the code?

hageboeck replied:
It looks historical, because it says in the comment that it's deep, but it's not.
```c++
NumericalDerivator(const NumericalDerivator &other) = default;
```
?

egpbos replied:
This was indeed historical (there used to be two copy ctors, and at that point the transformers were probably not part of the class yet). By the way, those trafo classes have no data members, so maybe they should just be free or static functions anyway, but that's outside of the scope of this PR.

guitargeek commented on math/minuit2/inc/Minuit2/NumericalDerivator.h:
```diff
@@ -0,0 +1,114 @@
+#ifndef ROOT_Minuit2_NumericalDerivator
+#define ROOT_Minuit2_NumericalDerivator
+
+#ifndef ROOT_Math_IFunctionfwd
+#include <Math/IFunctionfwd.h>
+#endif
```
```suggestion
#include <Math/IFunctionfwd.h>
```
The `Math/IFunctionfwd.h` file already has the include guard in it. Why repeating it here?

egpbos replied:
Yeah, that's so weird!

guitargeek commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+{
+   assert(function != nullptr && "function is a nullptr");
+
+   if (fVx.size() != function->NDim()) {
+      fVx.resize(function->NDim());
+   }
+   if (fVxExternal.size() != function->NDim()) {
+      fVxExternal.resize(function->NDim());
+   }
+   if (fVxFValCache.size() != function->NDim()) {
+      fVxFValCache.resize(function->NDim());
+   }
```
```suggestion
   fVx.resize(function->NDim());
   fVxExternal.resize(function->NDim());
   fVxFValCache.resize(function->NDim());
```
I think it's more readable without this size check, and the `std::vector<T>::resize()` method probably already knows how to handle the trivial case of no resizing.

egpbos replied:
Ah, that's a good point, hadn't thought of that. I guess it'd be implementation dependent, but indeed, makes the code cleaner and it was probably a premature optimization anyway.

guitargeek commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+
+      var2 = Ext2int(*parameter, sav2);
+      double vmin = var2 - var;
+      double gsmin = 8. * eps2 * (fabs(var) + eps2);
```
```suggestion
      double gsmin = 8. * eps2 * (std::abs(var) + eps2);
```
We should be consistent with the function for `abs`, and `std::abs` is preferred over `fabs` in C++ as far as I know (there are also other occurences of `fabs` that need to be replaced in this file).

egpbos replied:
`abs`olutely. I'll change them in one go then.

guitargeek commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+   }
+}
+
+bool NumericalDerivator::AlwaysExactlyMimicMinuit2() const
+{
+   return fAlwaysExactlyMimicMinuit2;
+};
+
+void NumericalDerivator::SetAlwaysExactlyMimicMinuit2(bool flag)
+{
+   fAlwaysExactlyMimicMinuit2 = flag;
+}
```
This can be inlined in the header.

guitargeek commented on math/minuit2/src/NumericalDerivator.cxx:
```diff
@@ -0,0 +1,336 @@
+}
+
+std::vector<DerivatorElement>
+NumericalDerivator::Differentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *cx,
```
```suggestion
NumericalDerivator::Differentiate(const ROOT::Math::IBaseFunctionMultiDim &function, const double *cx,
```
`function` can not be a `nullptr` right? In this case, the function parameters could be taken by reference to make the code clearer.

egpbos replied:
Ah, I have to check to be sure, but I think you're right. This was probably again a historical thing. There's even an assert checking for nullptr in SetupDifferentiate, so seems highly unlikely that nullptrs still make it through... Ah yeah, I think I remember now that I used to also have a copy of the function (pointer) in the class itself, but then for the newer class I needed to pass it in dynamically, so I added a check somewhere for nullptr calls, I think, that would use the internal function pointer...

Ok, so I think it can indeed be turned into a reference. However, I'm not completely confident that this won't mess up anything with the complicated inheritance hierarchy and pointer casting magic in #8596... Do you think it's ok to postpone this change?

guitargeek replied:
Sure, we can do it later indeed!

guitargeek reviewed (changes requested):
Hi, I have mostly minor comments, but also one more important question about the copy constructor.

egpbos commented:
I have applied all suggestions, thanks again for the reviews @hageboeck and @guitargeek. If you agree with delaying the pointer -> reference refactoring (https://github.com/root-project/root/pull/8567#discussion_r670458226), mostly because it would be a lot easier and faster to check once #8596 and the PR after that are merged, then I think this is ready to merge.

guitargeek reviewed (approved):
Thank you very much for the development of the `Numericalderivator` class for Minuit2 and addressing the comments by @lmoneta, @hageboeck, and me! The unaddressed comment with the references instead of pointers can be indeed addressed later. This PR is ready to be merged once the Jenkins tests pass.

egpbos commented:
The two failing Jenkins tests are not related to this PR, correct?

guitargeek commented:
No, the Jenkins failures are fortunately not related to this PR.

hahnjo commented:
FYI @guitargeek @egpbos this seems to break the default configuration because `minuit2` is off by default...

egpbos commented:
@hahnjo indeed, this is fixed in #8689. There could still be weird cases where RooFit is set to ON but Minuit2 set OFF manually, so that should be checked. We discussed this, but I forgot to implement it in this PR, sorry for that.
